File: Data Structures & Algorithms/count-number-of-islands/submission-2.java
class Solution {

    private static final int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

    public boolean isValid(int rows, int cols, int r, int c){
        return r >= 0 && r < rows && c >= 0 && c < cols; 
    }

    public void bfs(char[][] grid, int r, int c){
        grid[r][c] = '0';
        int rows = grid.length;
        int cols = grid[0].length;
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{r, c});
        while(!q.isEmpty()){

            int[] position = q.poll();
            r = position[0]; 
            c = position[1];
            
            for(int[] dir: directions){
                int nr = r + dir[0];
                int nc = c + dir[1];

                if(isValid(rows, cols, nr, nc) && grid[nr][nc] == '1'){
                    grid[nr][nc] = '0';
                    q.offer(new int[]{nr, nc});
                }
            }
        }
    }  

    public int numIslands(char[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;

        int numIslands = 0;
        for(int r = 0; r < rows; r++){
            for(int c = 0; c < cols; c++){
                if(grid[r][c] == '1'){
                    grid[r][c] = '0';
                    bfs(grid, r, c);
                    numIslands++;
                }
            }
        }

        return numIslands;
        
    }
}

File: Data Structures & Algorithms/surrounded-regions/submission-2.py
class Solution:
    def solve(self, grid: List[List[str]]) -> None:
        rows, cols = len(grid), len(grid[0])
        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]

        
        q = deque()
        for r in range(rows):
            if grid[r][0] == 'O':
                grid[r][0] = 'T'
                q.append((r, 0))
            if grid[r][-1] == 'O':
                grid[r][-1] = 'T'
                q.append((r, cols - 1))

        for c in range(cols):
            if grid[0][c] == 'O':
                grid[0][c] = 'T'
                q.append((0, c))
            if grid[-1][c] == 'O':
                grid[-1][c] = 'T'
                q.append((rows - 1, c))

        while q:
            r, c = q.popleft()
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if (
                    0 <= nr < rows 
                    and 0 <= nc < cols 
                    and grid[nr][nc] 
                    == 'O'
                ):
                    grid[nr][nc] = 'T'
                    q.append((nr, nc))
        
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 'O':
                    grid[r][c] = 'X'
                elif grid[r][c] == 'T':
                    grid[r][c] = 'O'



        